### PR TITLE
Close tabs on exit feature in Android

### PR DESCRIPTION
**Pull request by yrliou**
Fix https://github.com/brave/brave-browser/issues/4844

Note: This PR is based on https://github.com/brave/brave-core/pull/3427, will have to wait after it's merged.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Android
  - [x] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [x] iOS
  - [ ] Linux
  - [x] macOS
  - [x] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/BraveActivity.java [hunk 1]
```diff
@@ -32,5 +33,10 @@ public void onStartWithNative() {
         PrefServiceBridge.getInstance().setBoolean(Pref.NTP_ARTICLES_LIST_VISIBLE, false);
     }
 
+    public boolean isNoRestoreState() {
+        String PREF_CLOSE_TABS_ON_EXIT = "close_tabs_on_exit";
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/BraveActivity.java]: I would declare it on top of the class as `static final`

Comment (yrliou) [reply android/java/org/chromium/chrome/browser/BraveActivity.java]: Addressed in https://github.com/brave/brave-core/pull/3428/commits/b4d101eb6aa4a7e0a967e68d410f7c1ecd6550a3, thanks!

**Comment by yrliou** [comment]
Rebased and force pushed to run CI again.

**Comment by yrliou** [comment]
CI fail in network-audit on linux build, which should be unrelated to this change since this PR only touches android files.
```
15:32:33  + npm run network-audit -- --output_path=src/out/Release/brave
15:32:51  
15:32:51  > brave@0.72.26 network-audit /home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit
15:32:51  > node ./scripts/commands.js start --enable_brave_update --network_log --user_data_dir_name=brave-network-test "--output_path=src/out/Release/brave"
15:32:51  
15:32:51  Network audit started. Logging requests for the next 2min or until you quit Brave...
15:32:51  undefined: src/out/Release/brave --enable-logging --v=0 --enable_brave_update --network_log --user_data_dir_name=brave-network-test --output_path=src/out/Release/brave --user-data-dir=/home/ubuntu/.config/BraveSoftware/brave-network-test --log-net-log=/home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit/network_log.json --net-log-capture-mode=Everything
15:32:53  Xlib:  extension "XInputExtension" missing on display ":3".
15:32:53  Xlib:  extension "XInputExtension" missing on display ":3".
15:32:53  Xlib:  extension "XInputExtension" missing on display ":3".
15:32:53  Xlib:  extension "XInputExtension" missing on display ":3".
15:32:53  [28155:28155:0916/223253.113047:ERROR:gl_surface_glx.cc(460)] GLX 1.3 or later is required.
15:32:53  [28155:28155:0916/223253.113108:ERROR:gl_initializer_x11.cc(148)] GLSurfaceGLX::InitializeOneOff failed.
15:32:53  [28155:28155:0916/223253.114298:ERROR:viz_main_impl.cc(167)] Exiting GPU process due to errors during initialization
15:32:53  Xlib:  extension "XInputExtension" missing on display ":3".
15:32:53  [28129:28166:0916/223253.532475:ERROR:bus.cc(394)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
15:32:53  
15:32:53  (brave:28129): LIBDBUSMENU-GLIB-WARNING **: Unable to get session bus: Unknown or unsupported transport 'disabled' for address 'disabled:'
15:32:53  [28129:28169:0916/223253.704769:ERROR:rewards_service_impl.cc(196)] Failed to read file: /home/ubuntu/.config/BraveSoftware/brave-network-test/Default/ledger_state
15:32:53  [1:7:0916/223253.742945:ERROR:command_buffer_proxy_impl.cc(124)] ContextResult::kTransientFailure: Failed to send GpuChannelMsg_CreateCommandBuffer.
15:32:53  [28178:28178:0916/223253.743605:ERROR:ledger_impl.cc(100)] Failed to initialize wallet
15:35:00  undefined:17475
15:35:00  {"phase":2,"source":{"id":50,"type":8}]}
15:35:00                                        ^
15:35:00  
15:35:00  SyntaxError: Unexpected token ] in JSON at position 77909214
15:35:00      at JSON.parse (<anonymous>)
15:35:00      at start (/home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit/lib/start.js:144:23)
15:35:00      at Command.listener (/home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit/node_modules/commander/index.js:315:8)
15:35:00      at Command.emit (events.js:203:13)
15:35:00      at Command.parseArgs (/home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit/node_modules/commander/index.js:654:12)
15:35:00      at Command.parse (/home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit/node_modules/commander/index.js:474:21)
15:35:00      at Object.<anonymous> (/home/ubuntu/jenkins/workspace/wser-build-pr_close_tabs_on_exit/scripts/commands.js:155:4)
15:35:00      at Module._compile (internal/modules/cjs/loader.js:868:30)
15:35:00      at Object.Module._extensions..js (internal/modules/cjs/loader.js:879:10)
15:35:00      at Module.load (internal/modules/cjs/loader.js:731:32)
15:35:00  npm ERR! code ELIFECYCLE
15:35:00  npm ERR! errno 1
15:35:00  npm ERR! brave@0.72.26 network-audit: `node ./scripts/commands.js start --enable_brave_update --network_log --user_data_dir_name=brave-network-test "--output_path=src/out/Release/brave"`
15:35:00  npm ERR! Exit status 1
15:35:00  npm ERR! 
15:35:00  npm ERR! Failed at the brave@0.72.26 network-audit script.
15:35:00  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
15:35:00  
15:35:00  npm ERR! A complete log of this run can be found in:
15:35:00  npm ERR!     /home/ubuntu/.npm/_logs/2019-09-16T22_34_51_301Z-debug.log
```

**Comment by yrliou** [comment]
@bridiver Could you help review the patch change in this PR as the code owner? Thanks.

**Comment by yrliou** [comment]
Have to rebase again to resolve conflict in `android_brave_strings.grd`.

File: patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch [hunk 2]
```diff
@@ -11,3 +11,11 @@ index a168b085348c218c2951964beb37e0a25d754836..ad5cf672b9548a4f9921c4bc3521a0c8
      @IntDef({BackPressedResult.NOTHING_HAPPENED, BackPressedResult.HELP_URL_CLOSED,
              BackPressedResult.MINIMIZED_NO_TAB_CLOSED, BackPressedResult.MINIMIZED_TAB_CLOSED,
              BackPressedResult.TAB_CLOSED, BackPressedResult.TAB_IS_NULL,
+@@ -1161,6 +1161,7 @@ public class ChromeTabbedActivity
+                     CipherFactory.getInstance().restoreFromBundle(getSavedInstanceState());
+ 
+             boolean noRestoreState =
++                    super.isNoRestoreState() ||
```
Comment (bridiver) [inline patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch]: can't we just set this command line option?

Comment (yrliou) [reply patches/chrome-android-java-src-org-chromium-chrome-browser-ChromeTabbedActivity.java.patch]: You were right, addressed in https://github.com/brave/brave-core/pull/3428/commits/acb75f104578bcbc1bba91e752f88c12cf5a54ba.
No patching is required now, thank you!

**Pull request merged**